rmw_opendds_shared_cpp/src/OpenDDSNode.cpp

```diff
@@ -0,0 +1,241 @@
+}
+
+void OpenDDSNode::add_sub(const DDS::InstanceHandle_t& sub, const std::string& topic_name, const std::string& type_name)
+{
+  DDS::GUID_t part_guid = dpi_->get_repoid(dp_->get_instance_handle());
+  DDS::GUID_t guid = dpi_->get_repoid(sub);
+  sub_listener_->add_information(part_guid, guid, topic_name, type_name, EntityType::Subscriber);
+  sub_listener_->trigger_graph_guard_condition();
+}
+
+bool OpenDDSNode::remove_pub(const DDS::InstanceHandle_t& pub)
+{
+  DDS::GUID_t guid = dpi_->get_repoid(pub);
+  bool r = pub_listener_->remove_information(guid, EntityType::Publisher);
+  if (r) {
+    pub_listener_->trigger_graph_guard_condition();
+  } else {
+    RMW_SET_ERROR_MSG("pub_listener_->remove_information failed");
+  }
+  return r;
+}
+
+bool OpenDDSNode::remove_sub(const DDS::InstanceHandle_t& sub)
+{
+  DDS::GUID_t guid = dpi_->get_repoid(sub);
+  bool r = sub_listener_->remove_information(guid, EntityType::Subscriber);
+  if (r) {
+    sub_listener_->trigger_graph_guard_condition();
+  } else {
+    RMW_SET_ERROR_MSG("sub_listener_->remove_information failed");
+  }
+  return r;
+}
+
+rmw_ret_t OpenDDSNode::count_publishers(const char * topic_name, size_t * count)
+{
+  if (!topic_name) {
+    RMW_SET_ERROR_MSG("topic name is null");
+    return RMW_RET_ERROR;
+  }
+  if (!count) {
+    RMW_SET_ERROR_MSG("count is null");
+    return RMW_RET_ERROR;
+  }
+  *count = pub_listener_->count_topic(topic_name);
+  return RMW_RET_OK;
+}
+
+rmw_ret_t OpenDDSNode::count_subscribers(const char * topic_name, size_t * count)
+{
+  if (!topic_name) {
+    RMW_SET_ERROR_MSG("topic name is null");
+    return RMW_RET_ERROR;
+  }
+  if (!count) {
+    RMW_SET_ERROR_MSG("count is null");
+    return RMW_RET_ERROR;
+  }
+  *count = sub_listener_->count_topic(topic_name);
+  return RMW_RET_OK;
+}
+
+OpenDDSNode::OpenDDSNode(rmw_context_t & context)
+  : context_(context)
+  , gc_(nullptr)
+  , pub_listener_(nullptr)
+  , sub_listener_(nullptr)
+  , dp_()
+  , dpi_(nullptr)
+{
+  try {
+    gc_ = create_guard_condition(context_.implementation_identifier);
+    if (!gc_) {
+      throw std::runtime_error("create_guard_condition failed");
+    }
+    pub_listener_ = CustomPublisherListener::Raf::create(context_.implementation_identifier, gc_);
+    if (!pub_listener_) {
+      throw std::runtime_error("CustomPublisherListener failed");
+    }
+    sub_listener_ = CustomSubscriberListener::Raf::create(context_.implementation_identifier, gc_);
+    if (!sub_listener_) {
+      throw std::runtime_error("CustomSubscriberListener failed");
+    }
+    dp_ = context.impl->dpf_->create_participant(
+      static_cast<DDS::DomainId_t>(context.options.domain_id), PARTICIPANT_QOS_DEFAULT, 0, 0);
+    if (!dp_) {
+      throw std::runtime_error("create_participant failed");
+    }
+    dpi_ = dynamic_cast<OpenDDS::DCPS::DomainParticipantImpl*>(dp_.in());
+    if (!dpi_) {
+      throw std::runtime_error("casting to DomainParticipantImpl failed");
+    }
```

Comment:
Thought I saw a large portion of this is duplicated from earlier. Might be good to have a helper function to improve maintenance of duplicated blocks. Then again it could be my mistake.

Reply:
Maybe you saw the same logic in constructing OpenDDSStaticPublisherInfo, but they are creating different things, no duplicate. One thing I could simplify is to take out context_.implementation_identifier. Since that will affect three classes and we want this PR to be merged first, I decided to do that in next PR.